pkg/resource/deploy/plan_executor.go

```diff
@@ -148,19 +145,35 @@ func (pe *planExecutor) Execute(callerCtx context.Context, opts Options, preview
 					// Signal completion to the step executor. It'll exit once it's done retiring all of the steps in
 					// the chain that we just gave it.
 					pe.stepExec.SignalCompletion()
-					log.Infof("planExecutor.Execute(...): issued deletes")
+					logging.V(4).Infof("planExecutor.Execute(...): issued deletes")
 
 					return false, nil
 				}
 
-				if eventErr := pe.handleSingleEvent(event.Event); eventErr != nil {
-					log.Infof("planExecutor.Execute(...): error handling event: %v", eventErr)
-					pe.reportError(pe.plan.generateEventURN(event.Event), eventErr)
+				flow, eventErr := pe.handleSingleEvent(event.Event)
+				if eventErr != nil || flow == Bail {
+					if eventErr != nil {
+						logging.V(4).Infof("planExecutor.Execute(...): error handling event: %v", eventErr)
+						pe.reportError(pe.plan.generateEventURN(event.Event), eventErr)
+					}
+
+					logging.V(4).Infof("planExecutor.Execute(...): canceling context")
 					cancel()
+
+					// TODO(swgillespie) - As we plumb new error handling code throughout the engine, we will
+					// be able to propegate the `Bail` upwards to our callers so that we don't need to fabricate
+					// fake errors here to force the process to tear down.
+					//
+					// For now, creating a bogus error is sufficient to cause this function to emit a "plan faliled"
+					// error that ultimately bubbles up to the top level.
+					if eventErr == nil {
+						eventErr = errors.New("step generation failed")
+					}
```

Comment:
this definitely seems like a reasonable way to 'snip' off the virality of this change at certain points.  However, in general, it would be nice if there was some type/pattern we could use here so you can go back later and move these forward.  i.e. have some sort of helper function (which simply makes an error) but now can be documetned and can be something we look for and remove later.

Reply:
Something like `context.TODO()` is for contexts? That would make sense here for sure